### DISCUSSION TITLE
Radiolib LoraWAN Helium network compatibility

### DISCUSSION THREAD

**Discussion by vallentb** [jgromes/RadioLib#932] [General]
Hi guys, I played with Radiolib on Helium network, but seems there is an issue with it. On first boot it works, but rejoiun not possible. Could you check and confirm? MTX in advance

**Comment by StevenCellist** [comment]
Hi @vallentb, I cannot confirm as I don't use Helium. So you'll need to provide information for us to work with. A few points to help you along:

* Does your Helium server support LoRaWAN v1.1? RadioLib really works best on v1.1 and is likely to not adhere to 1.0.x specifications at some details.
* Please be very complete on the following items:
* What board do you use?
* What does your Helium console show? Try to get as much debug/verbose output as possible.
* What does your board show as debug output? You can enable debug output in BuildOptUser.h file.

**Comment by vallentb** [comment]
Dear Steven,
First of all I would like to thanks for your fast response!

According to Helium network documentation: The Helium Network supports any LoRaWAN capable device meeting the v1.0.2, v1.0.3, or v1.0.4 specification.
https://docs.helium.com/iot/devices/

I use Heltec Wireless tracker in my test:
https://heltec.org/project/wireless-tracker/

As I wrote if I fully erase the board with ESP tools, set up new device in Helium console (Device EUI, App EUI, App Key, only OTAA activation available on Helium network), the first join session wurks, I see valid uplinks, acknowledge from Helium. But if I restart the board, never had join accept, I can not see Join request in the console. The debug output ending with  failed, code -6. According to your examples you clearly state that  "REQUIRES you to check "Resets DevNonces" on your LoRaWAN dashboard.", but this is not possible on the Helium console, just  TTN provide this option Console->end devices (general setting, reset join nonces). 
Here is my code what I uploaded on my board

`#include <RadioLib.h>
// Module(NSS/CS, DIO1, RESET, BUSY)
SX1262 radio = new Module(8, 14, 12, 13);
LoRaWANNode node(&radio, &EU868);


void setup() {
  Serial.begin(9600);
  // initialize SX1278 with default settings
  Serial.print(F("[SX1278] Initializing ... "));
  int state = radio.begin();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }

  uint64_t joinEUI = 0x0000000000000000; // app eui
  uint64_t devEUI = 0x6081F9B0ACB9E42F;
  uint8_t nwkKey[] = { 0xFF, 0x84, 0x05, 0x17, 0x04, 0x3B, 0x20, 0x05, 0x05, 0x67, 0x72, 0x95, 0x80, 0xF5, 0xE4, 0x6F };
  //uint8_t nwkKey[] = {NULL};
  uint8_t appKey[] = {NULL};
  //uint8_t appKey[] = { 0xFF, 0x84, 0x05, 0x17, 0x04, 0x3B, 0x20, 0x05, 0x05, 0x67, 0x72, 0x95, 0x80, 0xF5, 0xE4, 0x6F };
  
  node.wipe();
  uint8_t joinDr = 4;
  
  Serial.print(F("[LoRaWAN] Attempting over-the-air activation ... "));
  state = node.beginOTAA(joinEUI, devEUI, nwkKey, appKey, joinDr);

  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }

/*
  // on EEPROM-enabled boards, after the device has been activated,
  // the session can be restored without rejoining after device power cycle
  // this is intrinsically done when calling `beginOTAA()` with the same keys
  // or if you 'lost' the keys or don't want them included in your sketch
  // you can call `restore()`
  
    Serial.print(F("[LoRaWAN] Resuming previous session ... "));
    state = node.restore();
    if(state == RADIOLIB_ERR_NONE) {
      Serial.println(F("success!"));
    } else {
      Serial.print(F("failed, code "));
      Serial.println(state);
      while(true);
    }
*/
  // disable the ADR algorithm
  node.setADR(false);

  // set a fixed datarate
  node.setDatarate(4, true);

  // enable CSMA
  // this tries to minimize packet loss by searching for a free channel
  // before actually sending an uplink
  // node.setCSMA(6, 2, true);

  // enable or disable the dutycycle
  // the second argument specific allowed airtime per hour in milliseconds
  // 1250 = TTN FUP (30 seconds / 24 hours)
  // if not called, this corresponds to setDutyCycle(true, 0)
  // setting this to 0 corresponds to the band's maximum allowed dutycycle by law
  //node.setDutyCycle(true, 1250);

  // enable or disable the dwell time limits
  // the second argument specific allowed airtime per uplink in milliseconds
  // if not called, this corresponds to setDwellTime(true, 0)
  // setting this to 0 corresponds to the band's maximum allowed dwell time by law
  //node.setDwellTime(true, 1000);
}

void loop() {
  int state = RADIOLIB_ERR_NONE;

  // set battery fill level - the LoRaWAN network server
  // may periodically request this information
  // 0 = external power source
  // 1 = lowest (empty battery)
  // 254 = highest (full battery)
  // 255 = unable to measure
  uint8_t battLevel = 146;
  node.setDeviceStatus(battLevel);

  // retrieve the last uplink frame counter
  uint32_t fcntUp = node.getFcntUp();

  Serial.print(F("[LoRaWAN] Sending uplink packet ... "));
  String strUp = "Hello World! #" + String(fcntUp);

  // send a confirmed uplink to port 10 every 64th frame
  // and also request the LinkCheck and DeviceTime MAC commands
  if (fcntUp % 64 == 0) {
    state = node.uplink(strUp, 10, true);
    node.sendMacCommandReq(RADIOLIB_LORAWAN_MAC_LINK_CHECK);
    node.sendMacCommandReq(RADIOLIB_LORAWAN_MAC_DEVICE_TIME);
  } else {
    state = node.uplink(strUp, 10);
  }
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
  }

  // after uplink, you can call downlink(),
  // to receive any possible reply from the server
  // this function must be called within a few seconds
  // after uplink to receive the downlink!
  Serial.print(F("[LoRaWAN] Waiting for downlink ... "));
  String strDown;

  // you can also retrieve additional information about
  // uplink or downlink by passing a reference to
  // LoRaWANEvent_t structure
  LoRaWANEvent_t event;
  state = node.downlink(strDown, &event);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));

    // print data of the packet (if there are any)
    Serial.print(F("[LoRaWAN] Data:\t\t"));
    if (strDown.length() > 0) {
      Serial.println(strDown);
    } else {
      Serial.println(F("<MAC commands only>"));
    }

    // print RSSI (Received Signal Strength Indicator)
    Serial.print(F("[LoRaWAN] RSSI:\t\t"));
    Serial.print(radio.getRSSI());
    Serial.println(F(" dBm"));

    // print SNR (Signal-to-Noise Ratio)
    Serial.print(F("[LoRaWAN] SNR:\t\t"));
    Serial.print(radio.getSNR());
    Serial.println(F(" dB"));

    // print frequency error
    Serial.print(F("[LoRaWAN] Frequency error:\t"));
    Serial.print(radio.getFrequencyError());
    Serial.println(F(" Hz"));

    // print extra information about the event
    Serial.println(F("[LoRaWAN] Event information:"));
    Serial.print(F("[LoRaWAN] Direction:\t"));
    if (event.dir == RADIOLIB_LORAWAN_CHANNEL_DIR_UPLINK) {
      Serial.println(F("uplink"));
    } else {
      Serial.println(F("downlink"));
    }
    Serial.print(F("[LoRaWAN] Confirmed:\t"));
    Serial.println(event.confirmed);
    Serial.print(F("[LoRaWAN] Confirming:\t"));
    Serial.println(event.confirming);
    Serial.print(F("[LoRaWAN] Datarate:\t"));
    Serial.print(event.datarate);
    Serial.print(F("[LoRaWAN] Frequency:\t"));
    Serial.print(event.freq, 3);
    Serial.println(F(" MHz"));
    Serial.print(F("[LoRaWAN] Output power:\t"));
    Serial.print(event.power);
    Serial.println(F(" dBm"));
    Serial.print(F("[LoRaWAN] Frame count:\t"));
    Serial.println(event.fcnt);
    Serial.print(F("[LoRaWAN] Port:\t\t"));
    Serial.println(event.port);

    Serial.print(radio.getFrequencyError());

    uint8_t margin = 0;
    uint8_t gwCnt = 0;
    if (node.getMacLinkCheckAns(&margin, &gwCnt)) {
      Serial.print(F("[LoRaWAN] LinkCheck margin:\t"));
      Serial.println(margin);
      Serial.print(F("[LoRaWAN] LinkCheck count:\t"));
      Serial.println(gwCnt);
    }

    uint32_t networkTime = 0;
    uint8_t fracSecond = 0;
    if (node.getMacDeviceTimeAns(&networkTime, &fracSecond, true)) {
      Serial.print(F("[LoRaWAN] DeviceTime Unix:\t"));
      Serial.println(networkTime);
      Serial.print(F("[LoRaWAN] LinkCheck second:\t1/"));
      Serial.println(fracSecond);
    }

  } else if (state == RADIOLIB_ERR_RX_TIMEOUT) {
    Serial.println(F("timeout!"));

  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
  }

  // on EEPROM enabled boards, you can save the current session
  // by calling "saveSession" which allows retrieving the session after reboot or deepsleep
  
    //node.saveSession();
  

  // wait before sending another packet
  uint32_t minimumDelay = 60000;                  // try to send once every minute
  uint32_t interval = node.timeUntilUplink();     // calculate minimum duty cycle delay (per law!)
  uint32_t delayMs = max(interval, minimumDelay); // cannot send faster than duty cycle allows

  delay(delayMs);
}`

Thank your support in advance!

[debug output.odt](https://github.com/jgromes/RadioLib/files/13980986/debug.output.odt)

**Comment by StevenCellist** [comment]
_"According to your examples you clearly state that "REQUIRES you to check "Resets DevNonces" on your LoRaWAN dashboard.", but this is not possible on the Helium console, just TTN provide this option Console->end devices (general setting, reset join nonces)."_

Well, combining this statement with _"v1.0.2, v1.0.3, or v1.0.4"_ pretty much says enough - Helium is getting outdated and cannot be used on newer implementations (specifically, RadioLib). 

One point worth mentioning that could help you slightly: RadioLib does implement session persistence which only works if you do _not_ call `node.wipe()`. In that case, after a successful join, you can keep using this session even after power-down and/or reboot. This does require that you call `node.saveSession()` before any reboot / powercycle (which may end up requiring that you call it after every downlink, just to be sure, but that depends on your application). Thus, you only need to join once and can maybe get away with an older version (beware that it might still cause undefined behaviour, causing many many downlinks!).
Unfortunately, I cannot or will not be able to provide any other support for Helium.

**Reply by vallentb** [reply]
MTX Steven, I try the provided remedy. I hope Helium will update the network to the latest version ASAP.

**Comment by HeadBoffin** [comment]
From the first post it appears likely that this is a DevNonce issue - first join will work OK but further dev work resulting in loss of session or the need to wipe the session resets it which the LNS remembers. So the "Resets DevNonces" is for the purposes of convenience during development and shouldn't be used for production devices. I'd be surprised if this option wasn't available somewhere in the Helium LNS but I've rather lost track with the whole host your own console thing. At worst you can persist the DevNonce only or use ABP. 

But fundamentally, this library is for LW v1.1.0 so YMMV with the closest match of v1.0.4

**Reply by vallentb** [reply]
ThanksBoffin, ABP not available on Helium, wipe session is not available in the console :(

**Comment by tve** [comment]
This is an old thread but I just wanted to follow-up that I'm using Helium with the public'ly available Meteo Scientific LNS, which is really ChirpStack and there's a DevNonce reset button on the device's OTAA tab. So I'm not having any issues when the device restarts with DevNonce==0 due to whatever.

**Reply by clavisound** [reply]
@tve On another LNS, I only have join accept but the uplinks are ignored. You also have uplinks?

**Reply by StevenCellist** [reply]
@clavisound do you use LoRaWAN v1.1 on a Helium network? That does not work unfortunately. For Helium, you will have to stick to 1.0.4.

**Reply by clavisound** [reply]
@StevenCellist Yes. The join is accepted, but not the uplink. TTN is fine. RadioLib supports only 1.1, right? I don't understand how @tve managed to work with helium.

**Reply by StevenCellist** [reply]
If you had taken the time to read the README in the examples/LoRaWAN folder, you'd have seen that RadioLib supports 1.0.4 just the same.

**Reply by clavisound** [reply]
@StevenCellist I have missed that, thank you.

**Reply by tve** [reply]
Just to confirm: join request & accept, (un&)confirmed link up&down, ADR all work for me. I am using 1.0.4 (no NwkKey).

**Reply by clavisound** [reply]
@StevenCellist just returned to home to re-check and the examples does not link to the README, or not comment about the 1.0.4 version with the absence of nwkKey. More importantly on the [wiki](https://github.com/jgromes/RadioLib/wiki/LoRaWAN:-Device-setup-on-TTN) it states: "Choose LoRaWAN 1.1.0 - the last one in the list - the latest specfication. RadioLib uses Regional Parameters 1.1 revision B.", hence I missed the possibility of 1.0.4. Maybe it's better to add that info in the example or in wiki?

On purpose I tried to use the library as a user, not as developer. Installed from the library manager, so I missed the README on examples.

Just tried with Helium *and* TTN with 1.0.4 and I have mismatch on MIC on both networks. I substituted `nwkKey` with `NULL` on `beginOTAA`. I re-started Arduino to get-rid of caches.

Here the log. There is some question marks. Mind you, even with question marks LoRaWAN 1.1 joins and uplinks with TTN. This is the 1.0.4 log
```
19:17:58.491 -> 
19:17:58.491 -> Setup ... 
19:18:04.502 -> Initialise the radio
19:18:04.535 -> RLB_DBG: 
19:18:04.535 -> RadioLib Info
19:18:04.569 -> Version:  "7.1.2.0"
19:18:04.569 -> Platform: "Arduino AVR"
19:18:04.602 -> Compiled: "Mar 19 2025" "19:03:43"
19:18:04.635 -> RLB_DBG: M	SX127x
19:18:04.635 -> Join ('login') the LoRaWAN Network
19:18:04.701 -> RLB_PRO: Setting up dynamic channels
19:18:04.734 -> RLB_PRO: UL:   0 1       ? (0 - 5) | DL:   0 1       ? (0 - 5)
19:18:04.801 -> RLB_PRO: UL:   1 1       ? (0 - 5) | DL:   1 1       ? (0 - 5)
19:18:04.867 -> RLB_PRO: UL:   2 1       ? (0 - 5) | DL:   2 1       ? (0 - 5)
19:18:04.933 -> RLB_PRO: [MAC] 0x03
19:18:04.967 -> RLB_PRO: 00000000: 50                                               P                 
19:18:05.033 -> RLB_PRO: LinkAdrReq: dataRate = 5, txSteps = 0, nbTrans = 0
19:18:05.099 -> RLB_PRO: LinkAdrAns: 07
19:18:05.133 -> RLB_PRO: [MAC] 0x04
19:18:05.133 -> RLB_PRO: 00000000: 07                                               .                 
19:18:05.232 -> RLB_PRO: DutyCycleReq: max duty cycle = 1/2^7
19:18:05.298 -> RLB_PRO: [MAC] 0x05
19:18:05.298 -> RLB_PRO: 00000000: 00 d2 ad 84                                      ....              
19:18:05.398 -> RLB_PRO: RXParamSetupReq: Rx1DrOffset = 0, rx2DataRate = 0, freq =       ?
19:18:05.497 -> RLB_PRO: [MAC] 0x08
19:18:05.497 -> RLB_PRO: 00000000: 01                                               .                 
19:18:05.597 -> RLB_PRO: RXTimingSetupReq: delay = 1 sec
19:18:05.630 -> RLB_PRO: [MAC] 0x09
19:18:05.663 -> RLB_PRO: 00000000: 05                                               .                 
19:18:05.730 -> RLB_PRO: [MAC] 0x0c
19:18:05.763 -> RLB_PRO: 00000000: 65                                               e                 
19:18:05.862 -> RLB_PRO: ADRParamSetupReq: limitExp = 6, delayExp = 5
19:18:05.929 -> RLB_PRO: [MAC] 0x0f
19:18:05.929 -> RLB_PRO: 00000000: fa                                               .                 
19:18:06.028 -> RLB_PRO: RejoinParamSetupReq: maxTime = 15, maxCount = 10
19:18:06.095 -> RLB_PRO: 
19:18:06.095 -> RLB_PRO: PHY:  Frequency =       ? MHz, TX = 16 dBm
19:18:06.161 -> RLB_PRO: LoRa: SF = 7, BW =     ? kHz, CR = 4/5, IQ: U
19:18:06.194 -> RLB_DBG: Timeout in 93 ms
19:18:06.228 -> RLB_PRO: JoinRequest sent (DevNonce = 0) <-- Rx Delay start
ERASED
19:18:06.460 -> RLB_PRO: 
19:18:06.493 -> RLB_PRO: PHY:  Frequency =       ? MHz, TX = 16 dBm
19:18:06.526 -> RLB_PRO: LoRa: SF = 7, BW =     ? kHz, CR = 4/5, IQ: D
19:18:11.241 -> RLB_PRO: Opening Rx1 window (51 ms timeout)... <-- Rx Delay end 
19:18:11.274 -> RLB_PRO: Closing Rx1 window
19:18:11.307 -> RLB_PRO: 
19:18:11.341 -> RLB_PRO: PHY:  Frequency =       ? MHz, TX = 16 dBm
19:18:11.374 -> RLB_PRO: LoRa: SF = 12, BW =     ? kHz, CR = 4/5, IQ: D
19:18:12.204 -> RLB_PRO: Opening Rx2 window (689 ms timeout)... <-- Rx Delay end 
19:18:12.935 -> RLB_PRO: Closing Rx2 window
19:18:12.935 -> Join failed - RADIOLIB_ERR_NO_JOIN_ACCEPT (-1116)
```

**Reply by StevenCellist** [reply]
> the examples does not link to the README

Meh, will resolve.

> ERASED

No point erasing - any gateway around you is able to pick up your DevEUI and JoinEUI in plain text anyway.

> There is some question marks.

This is the core you use not properly being able to use printf-placeholders for floats.

> I have mismatch on MIC on both networks.

That is really up to your AppKey being incorrect(ly formatted), no other option. Sorry. It's just a plain hex array in the default order, no byte reversal, nothing. Simply copy-paste as hex array.

**Comment by HeadBoffin** [comment]
> So I'm not having any issues when the device restarts with DevNonce==0 due to whatever.

A full implementation of RadioLib with join persistence does not need this - this issue is only relevant to the development cycle if your storage is completely erased when you flash the device.
